INT-3885: Fix JMS Outbound Gateway Concurrency

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3885

Possible dropped reply, causing timeout.

WARN org.springframework.integration.jms.JmsOutboundGateway#1.replyListener-1 jms.JmsOutboundGateway:1202
    - Failed to consume reply with correlationId 164a49bf-c41d-4c0b-b012-55deecf001d1_2
      java.lang.RuntimeException: No sender waiting for reply
- Reproduced by running the test in a loop
- Cleaned up test to aid debugging - capture a unique message at each stage
- Added additional debug logging to th gateway

**cherry-pick to all**

garyrussell committed:
INT-3885: Fix JMS Outbound Gateway Concurrency

JIRA: https://jira.spring.io/browse/INT-3885

Possible dropped reply, causing timeout.

WARN org.springframework.integration.jms.JmsOutboundGateway#1.replyListener-1 jms.JmsOutboundGateway:1202
    - Failed to consume reply with correlationId 164a49bf-c41d-4c0b-b012-55deecf001d1_2
      java.lang.RuntimeException: No sender waiting for reply

- Reproduced by running the test in a loop
- Cleaned up test to aid debugging - capture a unique message at each stage
- Added additional debug logging to th gateway

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java:
```diff
@@ -1178,6 +1180,12 @@ public void onMessage(javax.jms.Message message) {
 			LinkedBlockingQueue<javax.jms.Message> queue = this.replies.get(correlationId);
 			if (queue == null) {
 				if (this.correlationKey != null) {
+					Log debugLogger = LogFactory.getLog("si.jmsgateway.debug");
```
???
Can't we rely on the normal supplied `logger` from the super `IntegrationObjectSupport` ?

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java:
```diff
@@ -1178,6 +1180,12 @@ public void onMessage(javax.jms.Message message) {
 				if (this.correlationKey != null) {
+					Log debugLogger = LogFactory.getLog("si.jmsgateway.debug");
+					if (debugLogger.isDebugEnabled()) {
+						Object siMessage = this.messageConverter.fromMessage(message);
```
OK. I think I got it. This is the point do not not `debug` with the parent preferences.

artembilan commented:
I think I'm fine with the fix.
Take a look one more time after polling it locally.

artembilan commented:
Well, looks like work hasn't been finished:
```
17:36:25,408  INFO main request_reply.PipelineJmsTests:203 - Test config: pipeline-named-queue-04.xml
17:36:25,408  INFO main request_reply.PipelineJmsTests:204 - Success: 49
17:36:25,408  INFO main request_reply.PipelineJmsTests:205 - Timeout: 1
17:36:25,408  INFO main request_reply.PipelineJmsTests:206 - Failure: 0

org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.springframework.integration.jms.request_reply.PipelineNamedReplyQueuesJmsTests$Capture] is defined
```
I can add `Capture` definition to all those contexts on merge, of course, but I wonder if it is a bell for you...

artembilan commented:
Maybe we should `synchronized (this.replies)` everywhere ?..

garyrussell committed:
Polishing

garyrussell commented:
Rebased, pushed.

I don't believe it's necessary to synchronize on `this.replies`; it's a `ConcurrentHashMap` and we only put/get.

My original plan here was to add diagnostics only; but I reproduced the problem on test 3a and fixed it.

I don't think it's necessary to add diagnostics to all the tests, although I have done so for test4 because you saw a failure.

I don't really like these tests; they were originally indeterminate (tests like "at least 10 successes" with random delays and some timeouts expected).

Over the years, these timeouts have been increased such that, in general, I don't think we should see any timeouts.

So, I propose we merge and observe the CI builds.

artembilan commented:
Agree. It's so long story to fight with these tests...
So, merging...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/f1bd6e3bac07e894a4d0910a58a79ddc5b3015d3 after squash.
